Added API tests to ReductionOpInfo and ported more tests

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63978
* #62900
* __->__ #62899
* #62737

Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

heitorschueroff committed:
Added API tests to ReductionOpInfo

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo"




[ghstack-poisoned]

mruberry commented on test/test_reductions.py:
```diff
@@ -55,19 +56,191 @@ def _rand_shape(dim, min_size, max_size):
         shape.append(random.randint(min_size, max_size))
     return tuple(shape)
 
+def reduced_shape(shape, dim=None, keepdim=False):
```
Add a comment for what this function does

mruberry commented on test/test_reductions.py:
```diff
@@ -55,19 +56,191 @@ def _rand_shape(dim, min_size, max_size):
+        self.assertEqual(op(t, *args, dim=[0, 2], keepdim=True, **kwargs).shape, [1, 3, 1])
+
+    @ops(filter(lambda op: not op.supports_multiple_dims, reduction_op_db), allowed_dtypes=[torch.float])
+    def test_dim_multi_unsupported(self, device, dtype, op: ReductionOpInfo):
```
Nice

mruberry commented on test/test_reductions.py:
```diff
@@ -55,19 +56,191 @@ def _rand_shape(dim, min_size, max_size):
+
+    @ops(reduction_op_db, allowed_dtypes=[torch.float])
+    def test_empty_slice(self, device, dtype, op: ReductionOpInfo):
+        """Tests for consistent behavior when reducing over an empty slice."""
```
Do we want a pointer to a doc or dev wiki or a comment here explaining what the empty slice reduction behavior should be? Or maybe just an elaborated comment here?

heitorschueroff replied:
I'll add more comments and a link to the GitHub issue discussing this

mruberry commented on test/test_reductions.py:
```diff
@@ -55,19 +56,191 @@ def _rand_shape(dim, min_size, max_size):
+
+    @ops(filter(lambda op: op.nan_policy == 'propagate', reduction_op_db),
+         allowed_dtypes=floating_and_complex_types_and(torch.bfloat16, torch.float16))
+    def test_nan_policy_propagate(self, device, dtype, op: ReductionOpInfo):
```
Cool tests

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8468,6 +8468,16 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
         supports_out=False,
         supports_forward_ad=True,
         dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+        skips=(
```
I'll add some more ops here.

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported amax/amin/nansum tests"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported amax/amin/nansum tests"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported amax/amin/nansum tests"




[ghstack-poisoned]

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
     return tuple(shape)
 
+def _reduced_shape(shape, dim=None, keepdim=False):
+    """Reduces the given dimensions of the input shape
```
More precisely this computes the expected reduced shape?

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
-        t = make_tensor((2, 3), device, dtype)
+    def _test_dim_keepdim(self, op: ReductionOpInfo, device, *, ndim, **dim_keepdim):
+        """Tests output shape for input with ndim and dim and keepdim kwargs"""
+        shape = torch.randint(2, 5, (ndim,)).tolist()
```
Using randint here isn't very interesting vs. just hand-writing some cases

mruberry replied:
(like 1D, 2D, and 3D are maybe the interesting cases?)

mruberry replied:
Nevermind - I was mistaken about the randint signature.

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+        self._test_dim_keepdim(op, device, ndim=3, dim=[0, 2], keepdim=True)
+
+    @ops(filter(lambda op: op.supports_multiple_dims, reduction_ops), dtypes=OpDTypes.none)
+    def test_dim_multi_unsorted(self, device, op: ReductionOpInfo):
```
Good idea to validate unsorted, too

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+            self._test_dim_keepdim(op, device, ndim=2, dim=2, keepdim=True)
+
+    @ops(filter(lambda op: op.identity is not None, reduction_ops), dtypes=OpDTypes.supported)
+    def test_identity(self, device, dtype, op: ReductionOpInfo):
```
This is a neat test

heitorschueroff replied:
Thanks!

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+        self.assertTrue(result.isnan())
+
+    @ops(filter(lambda op: op.nan_policy == 'omit', reduction_ops), dtypes=OpDTypes.supported,
+         allowed_dtypes=floating_and_complex_types_and(torch.bfloat16, torch.float16))
```
Nice reuse of the pattern from the identity tests (when the nan policy is omit)

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+        elif op.promotes_int_to_int64 and is_integral:
+            self.assertEqual(result.dtype, torch.int64)
+        else:
+            self.assertEqual(result.dtype, op.result_dtype or dtype)
```
This needs a comment for its unusual use of the or operator + the case that's being handled here

mruberry replied:
Alternatively this could add an additional clause for `op.result_dtype` being defined and then fallthrough to verify the dtype is preserved (that would probably be more readable)

heitorschueroff replied:
I'll add another elif for when result_dtype is not None and that shouldn't need comments which is better.

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+        result_with_nan = op(t, *args, **kwargs)
+        self.assertEqual(result, result_with_nan)
+
+    @ops(reduction_ops, dtypes=OpDTypes.supported)
```
Isn't the default for `dtypes` `OpDTypes.supported`?

heitorschueroff replied:
The default is OpDTypes.basic which only includes the default test dtypes. Here I want to test all supported dtypes.

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+              given inputs is floating point dtype.
+            - Otherwise, raise an error
+
+        See discussion here https://github.com/pytorch/pytorch/issues/61901
```
Nice summary + link

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+        The rules for reducing over an empty slice are as follows:
+            - Return the identity value if the operator has one
+            - Otherwise, return NaN if the operator promotes integral dtype to
+              floating point dtypes and the result of the operation for the
```
What's a case where the result of such an operation wouldn't be floating point?

heitorschueroff replied:
For operators that accept a dtype kwarg such as sum/prod. I'll add tests for `dtype` kwarg to ReductionOpInfo later and then update this test.

heitorschueroff replied:
However, now that I think about it I don't think we have any operator that promotes into to float that also has a dtype.

mruberry replied:
Setting the (hypothetical or otherwise) dtype kwarg of an operator that promotes int->float to an int dtype seems like an error, though, doesn't it?

heitorschueroff replied:
I don't think anyone would want to do that. NumPy allows it because it's part of the generic ufunc ops. So yeah, I think for now I'll remove this comment and assume there's never a case where an op whose codomain is floats would return an int.

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+            - Otherwise, return NaN if the operator promotes integral dtype to
+              floating point dtypes and the result of the operation for the
+              given inputs is floating point dtype.
+            - Otherwise, raise an error
```
It is a little interesting that an operation, like mode, will throw an error when given floating point inputs. 

An alternative rule might have been: if the result is a dtype that allows NaN, NaN, otherwise error.

heitorschueroff replied:
We could do that, but it would cause some inconsistencies. Like, `min` throws an error for integral inputs but returns `nan` for floating point inputs? In that case, why not return `inf` which is actually an identity for `min` but that only exists for floating point dtypes.

heitorschueroff replied:
I think it might be better to be more restrictive for now and sync with the Python Array API Standard on this. We can always stop raising an error for floating point results later.

mruberry commented on test/test_reductions.py:
```diff
@@ -55,18 +57,241 @@ def _rand_shape(dim, min_size, max_size):
+                    op(t, *args, dim=dim, **kwargs)
+
+    @ops(reduction_ops, dtypes=OpDTypes.none)
+    def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
```
Nice

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8733,13 +8699,64 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
             ),
         ),
     ),
+    ReductionOpInfo(
```
Same suggestion to insert these as alphabetically as possible

mruberry reviewed (approved):
A few suggestions/comments inline for your review but they all seem minor. Nice work

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported amax/amin/nansum tests"




[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408816).

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported amax/amin/nansum tests"


Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408816).

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported amax/amin/nansum tests"


Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408816).

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported more tests"


Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408816).

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported more tests"


Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408816).

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported more tests"


Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added API tests to ReductionOpInfo and ported more tests"


Differential Revision: [D30408816](https://our.internmc.facebook.com/intern/diff/D30408816)

[ghstack-poisoned]
